### DISCUSSION TITLE
Mesa Web

### DISCUSSION THREAD

**Discussion by Corvince** [mesa/mesa#1954] [Show and tell]
## Motivation

There have been talks in several places about hosting Mesa models on the web and also showcasing our example models similar to [Netlogo Web](https://www.netlogoweb.org/). One possible route is using [Jupyter Lite](https://jupyterlite.readthedocs.io/en/stable/) to statically host a Jupyter Lab and let the models run purely client-side. I have explored this direction and wanted to share first results, Here it is:

## Mesa Web
https://corvince.github.io/mesa-web

### Description
Currently I only included a slightly adjusted intro tutorial and the schelling model. But of course its easy to add more examples. You can even "upload" files yourself, but this will only persist on your own device. Otherwise it should work almost like a normal jupyter lab. You can install packages with `%pip install PACKAGE_NAME`, import modules (lots of modules are already included) and code your dream model. So its a nice interactive playground.

## Alternatives
I think @ankitk50 is working on something to host mesa examples solara apps side-by-side. I think this is also a promising direction and am very curious how it turns out. I think both approaches have some similarities, but also slightly different use-cases so I don't think they are competing. 

## Technical Description and Limitations
The basic setup was as easy as it gets: I followed [quickstart guide](https://jupyterlite.readthedocs.io/en/stable/quickstart/deploy.html) which uses a github template. This got me up and running a hosted jupyterlab instance in minutes. The difficult part was getting mesa installable since it indirectly depends on tornado, which is incompatible with pyodide. I solved it by providing a custom mesa wheel, which gets installed by `%pip install mesa`, so it doesn't install from PyPi. But this should be solved once we finished deprecating our old frontend. It is also possible to have mesa pre-installed so users can just import it. I'll have to look further into it to see how that is done. 
A big challenge is our solara frontend. I also got that working somewhat with a custom wheel (so you can install it), but its a bit more difficult to maintain and also doesn't work with mesa yet. I haven't fully looked into it, but it might be more difficult. That said I know the solara devs are also working on making solara more pyodide friendly, so lets see how that evolves.

**Comment by rht** [comment]
I can see this being used to replace https://mesa.readthedocs.io/en/stable/tutorials/intro_tutorial.html, as soon as applicable. As a first time reader, I find it slightly overwhelming to see the files, not knowing exactly whether I should create a new file from scratch, or click on the analysis.ipynb. But this issue is minor and can be addressed.

I think it should be possible to build static HTML output from the Sphinx build and have the doc entirely living on GH pages instead of RTD. The question is whether this is a better choice.

**Reply by rht** [reply]
It seems to make more sense to use Mesa Web only for showcasing the examples, given the IDE functionality of Jupyter Lite? No IDE is needed to read the tutorial, because it is meant to be read linearly.

**Reply by Corvince** [reply]
Good point, just found out that you can also create shareable links. So this one takes you straight to the tutorial notebook, without any unnecessary clutter
https://corvince.github.io/mesa-web/notebooks/index.html?path=intro_tutorial.ipynb

**Comment by tpike3** [comment]
Here is the Solara version Ankit is working on for examples, based @rlskoeser's example https://simrisk.pulcloud.io/

**Reply by wang-boyu** [reply]
We can probably have nested tabs: some top level tabs for example groups such as official, gis, community models, etc, and under these there're tabs for individual examples.

Mentioned in "Advanced tabs" section in https://solara.dev/api/tabs

**Comment by EwoutH** [comment]
As maybe relatively low-hanging fruit, can we add this to our whole example library? That would be totally amazing, if every examples was runnable directly in JupyterLite without modifying anything.

Two points:

- Are the app.py Solara files runnable somehow?
- What would be an elegant way to manage dependencies?
